### PR TITLE
add test for dataclusters

### PR DESCRIPTION
**Pull request by stevenhua0320**

### CODE REVIEW

**Review by sbillinge — COMMENTED** [review]
This is great, nice job! 

It might be cleaner if we define an __eq__ method in the class itself and then just assert that expected == actual. 

Also, possibly passing in a dict to the test and using this to instantiate the class may make it more readable. 

We are only testing one thing at the moment, which is fine, but we can decide if there are more edge cases we want to test. It can get time consuming so there is some balance between testing everything (better) and nothing (quicker in the short term).  

To think about this, remember we are testing "behavior" not code per se.  So putting text comments that capture what behavior each test is testing can be helpful.

**Comment by stevenhua0320** [comment]
> This is great, nice job!
> 
> It might be cleaner if we define an **eq** method in the class itself and then just assert that expected == actual.
> 
> Also, possibly passing in a dict to the test and using this to instantiate the class may make it more readable.
> 
> We are only testing one thing at the moment, which is fine, but we can decide if there are more edge cases we want to test. It can get time consuming so there is some balance between testing everything (better) and nothing (quicker in the short term).
> 
> To think about this, remember we are testing "behavior" not code per se. So putting text comments that capture what behavior each test is testing can be helpful.

Prof I'm not sure passing in a dict to the test means?  Is that mean I need to store all the attribute info into a dict and then instantiate the class by directly referencing the dict?

File: src/diffpy/srmise/dataclusters.py [hunk 1]
```diff
@@ -68,6 +68,21 @@ def __init__(self, x, y, res):
     def __iter__(self):
         return self
 
+    def __eq__(self, other):
+        # this function makes sure two DataClusters object is equal. Namely equal here means
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:72]: please remove comment

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+)
+def test_clear(inputs, expected):
+    # Initialize DataClusters with input parameters
+    actual = DataClusters(x=inputs["input_x"], y=inputs["input_y"], res=inputs["input_res"])
+
+    # Perform the clear operation
+    actual.clear()
+
+    # Assert each expected attribute against its actual value after clearing
+    for attr, expected_value in expected.items():
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:56]: this is no longer needed.  Just instantiate an expected with the outputs and then assert equal.  For readability, somthing like
```
input_cluster = DataClusters(x=inputs["x"]....)
expected = DataClusters(x=outputs["x"].....)
actual = input_cluster.clear()
assert expected == actual
```
If the clear operates in place and you can't assign to a new instance, then make the last line `assert expected == actual.clear`

**Review by sbillinge — COMMENTED** [review]
nice job.  Please see comment.s

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 2 above)
Comment (stevenhua0320) [reply src/diffpy/srmise/tests/test_dataclusters.py:56]: Initially I tried this approach but then I realized that we can't instantiate the expected object given those expected parameters. The reason why it cannot is because there is a function in the class called `setdata` that does not allow instantiating a object with 0 resolution as what clear does. here is part of the code that `setdata`defines:
`        if res <= 0:
            raise ValueError("Resolution res must be greater than 0.")`

Comment (stevenhua0320) [reply src/diffpy/srmise/tests/test_dataclusters.py:56]: The only way that makes your structure works is to refactor the source code here to change from `<=` to `<`, but I'm not sure if this would change the behavior of the function as Luke originally plans...

Comment (sbillinge) [reply src/diffpy/srmise/tests/test_dataclusters.py:56]: Why is it possible to set a zero resolution in clear but not when instantiating?  This is the question.   This is probably an issue with the code itself.  Actually, this is one reason we love tests, because we discover things like this....

Remember, tests test "behavior", not code.  What behavior do we want for the clear method?

Is clear used anywhere in the code?   Try and figure out what it is used for, then make a suggestion for your preferred fix.

File: src/diffpy/srmise/dataclusters.py [hunk 3]
```diff
@@ -68,6 +68,19 @@ def __init__(self, x, y, res):
...
+            and np.array_equal(self.clusters, other.clusters)
+            and self.res == other.res
+            and self.current_idx == other.current_idx
+            and self.lastcluster_idx == other.lastcluster_idx
+            and self.lastpoint_idx == other.lastpoint_idx
+            and self.status == other.status
+        )
+
     def clear(self):
         """Clear all members, including user data."""
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:85]: we need a better docstring.  When we touch code, we like to improve it.

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                "data_order": np.array([], dtype=np.int32),
+                "clusters": np.array([[]], dtype=np.int32),
+                "res": 0,
+                "current_idx": 0,
+                "lastcluster_idx": None,
+                "lastpoint_idx": None,
+                "status": 0,
+            },
+        ),
+        (
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:28]: what behavior is this testing that is different than the previous case?

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            {
+                "input_x": np.array([1, 2, 3]),
+                "input_y": np.array([3, 2, 1]),
+                "input_res": 4,
+            },
+            {
+                "x": np.array([]),
+                "y": np.array([]),
+                "data_order": np.array([], dtype=np.int32),
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:19]: why are we setting these to int32?

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    # Assert each expected attribute against its actual value after clearing
+    for attr, expected_value in expected.items():
+        assert (
+            np.array_equal(getattr(actual, attr), expected_value)
+            if isinstance(expected_value, np.ndarray)
+            else getattr(actual, attr) == expected_value
+        )
+
+
+@pytest.mark.parametrize(
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:64]: please put these in a different PR or we will never get this merged.  Each new test in a different PR

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    [
+        (
+            {
+                "input_x": np.array([1, 2, 3]),
+                "input_y": np.array([3, 2, 1]),
+                "input_res": 4,
+            },
+            DataClusters(np.array([1, 2, 3]), np.array([3, 2, 1]), 4),
+        ),
+        (
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:75]: what different behavior is this testing?

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                "input_x": np.array([1]),
+                "input_y": np.array([3]),
+                "input_res": 1,
+            },
+            DataClusters(np.array([1]), np.array([3]), 1),
+        ),
+    ],
+)
+def test_reset_clusters(inputs, expected):
+    actual = DataClusters(x=inputs["input_x"], y=inputs["input_y"], res=inputs["input_res"])
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:86]: please see comment above about naming these.

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    actual.reset_clusters()
+    assert actual == expected
+
+
+@pytest.mark.parametrize(
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:91]: please see comments above on this too.   Please start by doing a more informative docstring for each method you are testing.  This helps to understand what each method is doing.  It will also appear in the docs when we build them.

**Review by sbillinge — COMMENTED** [review]
please read all the inline comments.

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 2 above)
Comment (stevenhua0320) [reply src/diffpy/srmise/tests/test_dataclusters.py:56]: It is used in the initilization (__init__)of the object. Basically, it guarantees that it could create a clean object in the instantiating step. The code reference is here
`self.clear()`
`self.setdata(x, y, res)`

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 4 above)
Comment (stevenhua0320) [reply src/diffpy/srmise/tests/test_dataclusters.py:28]: Here I'm testing clearing object with multiple values in x & y lists, but since it is actually a method to reset everything into the very beginning state, I'm not sure if this case is necessary.

Comment (sbillinge) [reply src/diffpy/srmise/tests/test_dataclusters.py]: correct.  in both cases you are testing arrays.  That they are different lengths is kind of irrelevant.

I think for the clear function, we don't need to test too many different cases, if any.  For other functions we may want to test more different cases.

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 5 above)
Comment (sbillinge) [reply src/diffpy/srmise/tests/test_dataclusters.py:19]: yes, I kind of know why you did that, but I am also trying to emphasize a point that tests test behavior, not code, so we don't design tests to make code pass, we design tests to ensure specific behaviors, then we write code to make sure the tests pass.  Then we are sure that the code has the right behavior.  So understand why that is int32 and ask the question, why and is that the best thing?  We may not find the answer and then we can default to mimicking the original code, but we should at least go through that process.

File: src/diffpy/srmise/dataclusters.py [hunk 10]
```diff
@@ -82,7 +82,8 @@ def __eq__(self, other):
         )
 
     def clear(self):
-        """Clear all members, including user data."""
+        """Remove all the points from the cluster list,
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:85]: please write docstrings using numpy standard.

The first line is a short, high level description of what the method does, then a blank line, then more details, then any parameters and what is returned, then usage examples etc.  We often don't add usage examples or even long descriptions, so use your judgement.

The docstring is very important.  If you can't write a good docstring it means you don't understand what a method is for.  If you don't understand that you can't write good tests, and if you can't do that you can't write a good function.  So spend some effort on getting the docstring right first before doing other things, then return to the docstring later if your understanding changed while writing the tests.

**Review by sbillinge — COMMENTED** [review]
please see comments

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 2 above)
Comment (Ainamacar) [reply src/diffpy/srmise/tests/test_dataclusters.py:38]: Did not expect to see a flurry of activity on SrMise and thought I should chime in where it would be helpful.  A `res` of 0 should produce the trivial clustering -- each point is its own cluster.  I most likely required `res > 0` because the point of the class is to find non-trivial clusterings and because in an early stage of development this class was more tightly coupled to the physical resolution of the PDF, which certainly has a positive value.  Setting setting `res = 0` when clearing was a transient stand in for an uninitialized state.

Considered purely as a clustering method the trivial clustering is reasonable default behavior and the only change necessary is in `setdata()`, along the lines of
```
        if res < 0:
            raise ValueError("Resolution res must be non-negative.")
```

If you make that change, the docstring for `setvars()` in pdfpeakextraction.py should probably also be updated to state
"cres: The clustering resolution, must be >= 0".

Comment (sbillinge) [reply src/diffpy/srmise/tests/test_dataclusters.py:38]: Thanks Luke (@Ainamacar ) it is really great to have your input. I hope all is well with you!

We are trying to refactor all our diffpy packages to a standard format this summer to make maintenance easier moving forward. It is also great training for new people like Steven (@stevenhua0320 ).  Any input you can provide is really appreciated!

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 5 above)
Comment (Ainamacar) [reply src/diffpy/srmise/tests/test_dataclusters.py:19]: Both `data_order` and `clusters` store indices, which should be integers.  (It looks like I also did this with `check_qualidx` in modelcluster.py.)   So I believe this functions as a type hint or perhaps a vestige of habits from strongly typed languages.

Why int32?  I don't remember, but I can make an educated guess.  During original dev I used both 32 bit and 64 bit Python at various times and this might be a (very weak) gesture to compatibility between versions when serializing.  Alternately, sometimes I was getting different results on the same input and, IIRC, found that small numerical differences between the 32 bit and 64 bit versions occasional caused different in branching during the recursive steps of peak extraction.  So I may have specified int32 rather than int as a very early part of those debugging efforts, but never removed it.

I don't think changing this should have an impact -- if I knew about some subtle behavior I almost certainly would have left a comment.  My opinion is that it would be better to complete the conversion to Python3 and test on some full PDFs before seeing what happens if this is changed.  How that impacts what test should be written now, I'll leave to Simon.

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 2 above)
Comment (Ainamacar) [reply src/diffpy/srmise/tests/test_dataclusters.py:38]: Thank you, Simon!  I will email you a brief personal update.

In any case, I'd be happy to consult on this effort as I'm able.  I'm glad to see the whole diffpy effort is going strong!

File: src/diffpy/srmise/tests/test_dataclusters.py (hunk 5 above)
Comment (sbillinge) [reply src/diffpy/srmise/tests/test_dataclusters.py:19]: yes, please remove.  Per @Ainamacar this is technical debt and it generally better to remove it when we find it, than "trying to remember to remove it later if we need to".....

File: src/diffpy/srmise/dataclusters.py [hunk 11]
```diff
@@ -82,12 +82,28 @@ def __eq__(self, other):
         )
 
     def clear(self):
-        """Remove all the points from the cluster list,
-        and reset x, y value sequences to empty list. Reset the resolution to 0."""
+        """
+        Clear all data and reset the cluster object to a transient initial state.
+
+        This method performs the following steps:
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:88]: this block is too much information, so we can remove it.  It is information that is probably not needed by a user and a developer can look at the code to get the precise definition.

File: src/diffpy/srmise/dataclusters.py [hunk 12]
```diff
@@ -82,12 +82,28 @@ def __eq__(self, other):
...
+        - Remove all the points from the clusters list.
+        - Reset the x and y value sequences to empty NumPy arrays.
+        - Reset dataorder list to empty NumPy arrays.
+        - Reset the resolution to 0.
+        - Reset current_idx to 0.
+        - Reset lastcluster_idx and lastpoint_idx to None.
+        - Reset status to INIT.
+
+        The purpose of this method is to provide a clean state before creating new clustering operations.
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:97]: This is good, it is helpful.  I would also have a statement along the lines of "the object is updated in-place and no new instance is returned.", something like that, as this is something that I might like to know when I try and use it.

**Review by sbillinge — COMMENTED** [review]
Great job @stevenhua0320 this is looking great now.  One final comment.

**Review by sbillinge — COMMENTED** [review]
nearly there.  Please see inline comment.  We also want to finish up the changes that allow the `assert expected == actual` to pass as per the conversation.

**Comment by sbillinge** [comment]
> > nearly there. Please see inline comment. We also want to finish up the changes that allow the `assert expected == actual` to pass as per the conversation.
> 
> Simon, I tried the approach again, but it still seems that it could create error in this approach. The reason is that if we set expected for `x` and `y` be empty numpy list, then in the `setdata` function, it has two lines of code that correlates that. `self.data_order = self.y.argsort() # Defines order of clustering` `self.clusters = np.array([[self.data_order[-1], self.data_order[-1]]])` it means it could give an index of list based on y value. In this case, the list would be empty. But in the second line of code, since the cluster needs to take on value of the last index of value, it could cause `IndexOutOfBounds` error.

The test tests behavior and so we want the test to be what we want.  Then you need to modify code if the test is not passing.  You can't change the test to make things pass.

**Comment by sbillinge** [comment]
@stevenhua0320 to get the hand of this, why don't you write some tests for the eq method you wrote.  Since you write the function it will be easier to write tests. 

Normally we would write the tests before the function though....

**Comment by sbillinge** [comment]
Btw, please put that on a separate PR

File: src/diffpy/srmise/dataclusters.py [hunk 13]
```diff
@@ -106,21 +130,26 @@ def setdata(self, x, y, res):
...
-
-        self.data_order = self.y.argsort()  # Defines order of clustering
-        self.clusters = np.array([[self.data_order[-1], self.data_order[-1]]])
-        self.current_idx = len(self.data_order) - 1
-        self.lastcluster_idx = 0
-        self.lastpoint_idx = self.data_order[-1]
-
-        self.status = self.READY
+        # If x sequence size is empty, set the object into Initialized state.
+        if x.size == 0 and res == 0:
```
Comment (sbillinge) [inline src/diffpy/srmise/dataclusters.py:140]: what  will happen if `x.size == 0` but `res != 0`?

File: src/diffpy/srmise/tests/test_dataclusters.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
+import numpy as np
+import pytest
+
+from diffpy.srmise.dataclusters import DataClusters
+
+
+@pytest.mark.parametrize(
```
Comment (sbillinge) [inline src/diffpy/srmise/tests/test_dataclusters.py:7]: since we are only doing one test here, we don't need the mark.parameterize, you can just load the values directly into the test below.

**Review by sbillinge — COMMENTED** [review]
This is good. Please see inline comments.

I also have a comment about the eq method, but we can deal with that on the separate PR with the tests for that.

**Comment by stevenhua0320** [comment]
> This is good. Please see inline comments.
> 
> I also have a comment about the eq method, but we can deal with that on the separate PR with the tests for that.

Yeah I've seen that in the message. Since this is a method that is in the Datacluster class, I would like to push eq PR after this test is merged.

File: src/diffpy/srmise/dataclusters.py (hunk 13 above)
Comment (sbillinge) [reply src/diffpy/srmise/dataclusters.py:143]: This is not what I meant.  I think that we want the clear method test to pass (which it was) but if we were to write tests for this method, there would be other situations than `x.size == 0 and res == 0` and these were not being handled in your code fix.

Let's just revert this commit and I will merge, since it is passing tests, but write tests for this setdata method on a separate PR.

**Review by sbillinge — COMMENTED** [review]
please see comment

**Comment by sbillinge** [comment]
excellent @stevenhua0320 !
